src/tools/miri/src/machine.rs

```diff
@@ -1518,4 +1519,40 @@ impl<'tcx> Machine<'tcx> for MiriMachine<'tcx> {
             Entry::Occupied(oe) => Ok(oe.get().clone()),
         }
     }
+
+    fn get_global_alloc_salt(
+        ecx: &InterpCx<'tcx, Self>,
+        instance: Option<ty::Instance<'tcx>>,
+    ) -> usize {
+        let unique = if let Some(instance) = instance {
+            // Functions cannot be identified by pointers, as asm-equal functions can get deduplicated
+            // by the linker (we set the "unnamed_addr" attribute for LLVM) and functions can be
+            // duplicated across crates. We thus generate a new `AllocId` for every mention of a
+            // function. This means that `main as fn() == main as fn()` is false, while `let x = main as
+            // fn(); x == x` is true. However, as a quality-of-life feature it can be useful to identify
+            // certain functions uniquely, e.g. for backtraces. So we identify whether codegen will
+            // actually emit duplicate functions. It does that when they have non-lifetime generics, or
+            // when they can be inlined. All other functions are given a unique address.
+            // This is not a stable guarantee! The `inline` attribute is a hint and cannot be relied
+            // upon for anything. But if we don't do this, backtraces look terrible.
+            let is_generic = instance
+                .args
+                .into_iter()
+                .any(|kind| !matches!(kind.unpack(), ty::GenericArgKind::Lifetime(_)));
+            let can_be_inlined = match ecx.tcx.codegen_fn_attrs(instance.def_id()).inline {
+                InlineAttr::Never => false,
```

Comment:
https://github.com/rust-lang/rust/blob/60d146580c10036ce89e019422c6bc2fd9729b65/compiler/rustc_mir_transform/src/cross_crate_inline.rs#L17-L85 is the full logic that determines if a function may be inlined cross-crate. What you currently have is mostly an over-approximation of when functions may be inlined cross-crate, except that it misses the case where `tcx.sess.opts.unstable_opts.cross_crate_inline_threshold == InliningThreshold::Always` holds. In that case even `#[inline(never)]` can be inlined. (`#[inline(never)]` is still a hint, just a very strong hint)

Reply:
While the rest of your comment is accurate, 

> is the full logic that determines if a function may be inlined cross-crate. 

is not. The actual logic here is strewn all over the compiler, and I'm eventually going to try to bundle it all into this query.

But also, the output of this query depends on optimization level. So if all we do is invoke this query, you may actually get address instability in a real compilation that you don't see in Miri. Which would really look like a miscompile if you're not intimately familiar with this system.

Reply:
Overapproximating when cross-crate inlining may happen is fine. The current code in miri however currently misses the cross-crate inlining for `tcx.sess.opts.unstable_opts.cross_crate_inline_threshold == InliningThreshold::Always` though, which should be fixed IMO.

Reply:
Note that this logic is just being moved, this is already how Miri behaves. Also, if we wanted Miri to implement what we actually guarantee, we'd give *all* functions fresh addresses all the time. However, that would break backtraces... (specifically, the `__rust_begin_short_backtrace`/`__rust_end_short_backtrace` logic relies on comparing function pointers)